### ISSUE TITLE
basic_mixin can't be directly named as the target of an explicitly declared mixin typedef-name in a cursor

### ISSUE THREAD

**Issue by tahonermann** [ericniebler/stl2#268]
The following example code fails to compile:

```
#include <experimental/ranges/iterator>
using namespace std::experimental::ranges;
struct my_cursor {
  using mixin = basic_mixin<my_cursor>;
};
using my_iterator = basic_iterator<my_cursor>;
```

The reason is that <tt>basic_mixin</tt> is defined with a <tt>Destructible</tt> constraint that is not satisfied by an incomplete type, and <tt>my_cursor</tt> is incomplete at the point of the using declaration.

There is an easy "don't do that" solution - omit the mixin declaration and <tt>basic_iterator</tt> will use <tt>basic_mixin</tt> as specified above by default (and the type will be complete at the point that instantiation occurs).  It took me a little while to get to this solution from the reported errors though.

@CaseyCarter suggested the following as a possible solution via email:
> The better fix here would be to reduce the requirement on the base template from Destructible to "is object type," which I believe is satisfiable by incomplete types, and sufficient to guarantee Destructible in combination with the Constructible requirements on the constructors. (Since Constructible is a disjunction of "is an object type constructible from Args..." and "is a reference type that can be bound to Args..." it doesn't necessarily subsume Destructible.)

**Comment by CaseyCarter** [comment]
> The better fix here would be to reduce the requirement on the base template from Destructible to "is object type," which I believe is satisfiable by incomplete types, and sufficient to guarantee Destructible in combination with the Constructible requirements on the constructors. (Since Constructible is a disjunction of "is an object type constructible from Args..." and "is a reference type that can be bound to Args..." it doesn't necessarily subsume Destructible.)

Yes, Casey, `is_object` is satisfiable by incomplete object types. And `Constructible` will properly subsume `Destructible` after we land the wording changes in [P0547](wg21.link/p0547).
